### PR TITLE
feat(coderd/healthcheck): add access URL error codes and healthcheck doc

### PR DESCRIPTION
**Pull request by johnstcn**
Relates to https://github.com/coder/coder/issues/8965

- Added error codes for separate code paths in health checks
- Prefixed errors and warnings with error code prefixes
- Added a docs page with details on each code, cause and solution

Future work:
- add a convenience function in `site/src/api` to create a troubleshooting link from an error that extracts the error code prefix matching `^([A-Z0-9]+):`.
- Refactor both `.error` and `.warnings[]` to be a shape of `{ message: string, code: string}`. I started on this but realised it makes the diff too large, so will do it in a follow-up.

### CODE REVIEW

**Comment by BrunoQuaresma** [comment]
What would be the expected shape here? Something like:

```json
{
  "warnings": [
    "code": "W001",
    "message": "This is a warning"
  ],
  "resolutions": {
    "W001": "https://docs.coder.com/warning/W001"
  }
}
```

**Comment by johnstcn** [comment]
> What would be the expected shape here? Something like:
> 
> ```json
> {
>   "warnings": [
>     "code": "W001",
>     "message": "This is a warning"
>   ],
>   "resolutions": {
>     "W001": "https://docs.coder.com/warning/W001"
>   }
> }
> ```

I had envisioned `{ { code: string, message: string } ... }`, without the resolutions field.
Idea is that the FE would automatically construct this from `${CODER_DOCS_URL}/path/to/doc#W001`.

**Comment by BrunoQuaresma** [comment]
@johnstcn sounds good!

File: coderd/healthcheck/health/model.go [hunk 1]
```diff
@@ -1,9 +1,37 @@
...
+	CodeWebsocketEcho Code = "EWS02"
+	CodeWebsocketMsg  Code = "EWS03"
+
+	CodeAccessURLNotSet  Code = "EACS01"
+	CodeAccessURLInvalid Code = "EACS02"
+	CodeAccessURLFetch   Code = "EACS03"
+	CodeAccessURLNotOK   Code = "EACS04"
+
+	CodeDERPNodeUsesWebsocket Code = `EDERP01`
+	CodeDERPOneNodeUnhealthy  Code = `EDERP02`
```
Comment (johnstcn) [inline coderd/healthcheck/health/model.go:34]: review: these all have to go in here to be generated properly

File: docs/admin/healthcheck.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+Coder continuously executes a short database query to validate that it can reach its configured database, and also
+measures the median latency over 5 attempts.
+
+### <a name="EDB01">EDB01: Database Ping Failed</a>
+
+**Problem:** This error code is returned if any attempt to execute this database query fails.
+
+**Solution:** Investigate the health of the database.
+
+### <a name="EDB02">EDB02: Database Ping Slow</a>
```
Comment (matifali) [inline docs/admin/healthcheck.md:84]:
```suggestion
### <a name="EDB02">EDB02: Database Ping High</a>
```
I guess high/low are preferred over slow/fast for ping.

**Review by matifali — APPROVED** [review]
A few comments but this is Awesome <3

File: docs/admin/healthcheck.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+**Problem:** no access URL has been configured.
+
+**Solution:** configure an [access URL](./configure.md#access-url) for Coder.
+
+### <a name="EACS02">EACS02: Access URL invalid</a>
+
+**Problem:** `${CODER_ACCESS_URL}/healthz` is not a valid URL.
+
+**Solution:** Ensure that the access URL is a valid URL accepted by
+[`url.Parse`](https://pkg.go.dev/net/url#Parse).
```
[the remark this answers is not in the document]
Comment (matifali) [reply docs/admin/healthcheck.md:32]: Will work.

File: docs/admin/healthcheck.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+**Problem:** no access URL has been configured.
+
+**Solution:** configure an [access URL](./configure.md#access-url) for Coder.
+
+### <a name="EACS02">EACS02: Access URL invalid</a>
+
+**Problem:** `${CODER_ACCESS_URL}/healthz` is not a valid URL.
+
+**Solution:** Ensure that the access URL is a valid URL accepted by
```
Comment (mtojek) [inline docs/admin/healthcheck.md:31]: Maybe just paste a sample valid URL here (with scheme, trailing slashes, etc.)

File: docs/admin/healthcheck.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+[`url.Parse`](https://pkg.go.dev/net/url#Parse). You can check this [here](https://go.dev/play/p/CabcJZyTwt9).
+
+### <a name="EACS03">EACS03: Failed to fetch /healthz</a>
```
Comment (mtojek) [inline docs/admin/healthcheck.md:34]: nit: `/healthz`?

File: docs/admin/healthcheck.md [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+- A misconfigured reverse proxy
+- Invalid or expired SSL certificates
+
+**Solution:** Investigate and resolve the root cause of the connection issue.
+
+To troubleshoot further, you can log into the machine running Coder and attempt
+to run the following command:
+
+```shell
+curl -v ${CODER_ACCESS_URL}
```
Comment (mtojek) [inline docs/admin/healthcheck.md:52]: and `/healthz`

File: docs/admin/healthcheck.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+Coder continuously executes a short database query to validate that it can reach its configured database, and also
+measures the median latency over 5 attempts.
+
+### <a name="EDB01">EDB01: Database Ping Failed</a>
+
+**Problem:** This error code is returned if any attempt to execute this database query fails.
+
+**Solution:** Investigate the health of the database.
+
+### <a name="EDB02">EDB02: Database Ping High</a>
```
Comment (mtojek) [inline docs/admin/healthcheck.md:84]: Shouldn't be Latency High?

File: docs/admin/healthcheck.md [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+This may not be an error as such, but is an indication of a potential issue.
+
+**Solution:** Investigate the sizing of the configured database with regard to Coder's current activity and usage. It
+may be necessary to increase the resources allocated to Coder's database. Alternatively, you can raise the configured
+threshold to a higher value (this will not address the root cause).
+
+> [!TIP]
+> - You can enable [detailed database metrics](../cli/server.md#--prometheus-collect-db-metrics) in Coder's
+> Prometheus endpoint.
+> - Fif you have [tracing enabled](../cli/server.md#--trace), these traces may also contain useful information regarding
```
Comment (mtojek) [inline docs/admin/healthcheck.md:96]: nit: typo Fif

File: docs/admin/healthcheck.md [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+Coder workspace agents may use [DERP (Designated Encrypted Relay for Packets)](https://tailscale.com/blog/how-tailscale-works/#encrypted-tcp-relays-derp) to communicate with Coder.
+This requires connectivity to a number of configured [DERP servers](../cli/server.md#--derp-config-path) which are used
+to relay traffic between Coder and workspace agents. Coder periodically queries the health of its configured DERP servers and may return one or more of the following:
+
+### <a name="EDERP01">EDERP01: DERP Node Uses Websocket</a>
+
+**Problem:** When Coder attempts to establish a connection to one or more DERP servers, it sends a specific `Upgrade: derp` HTTP header.
+Some load balancers may block this header, in which case Coder will fall back to `Upgrade: websocket`.
+This is not necessarily a fatal error, but a possible indication of a misconfigured reverse HTTP proxy.
```
Comment (mtojek) [inline docs/admin/healthcheck.md:109]: Maybe add, that workspace users may be able to reach their workspaces, but connection performance might be degraded.

File: docs/admin/healthcheck.md [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+> [!NOTE]
+> This may also be shown if you have [forced websocket connections for DERP](../cli/server.md#--derp-force-websockets).
+
+**Solution:** ensure that any configured reverse proxy does not strip the `Upgrade: derp` header.
+
+### <a name="EDERP02">EDERP02: One or more DERP nodes unhealthy</a>
```
Comment (mtojek) [inline docs/admin/healthcheck.md:116]: nit: _is_ unhealthy ?

File: docs/admin/healthcheck.md [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+
+**Problem:** This s shown if Coder is unable to reach one or more configured DERP servers. Clients will fall back to use the remaining
```
Comment (mtojek) [inline docs/admin/healthcheck.md:118]: nit: This s

File: docs/admin/healthcheck.md [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+DERP servers, but performance may be impacted for clients closest to the unhealthy DERP server.
+
+**Solution:** Ensure that the DERP server is available and reachable over the network on port 443.
```
Comment (mtojek) [inline docs/admin/healthcheck.md:121]: Any curl here?

File: docs/admin/healthcheck.md [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+Coder makes heavy use of [WebSockets](https://datatracker.ietf.org/doc/rfc6455/) for long-lived connections:
+- Between users interacting with Coder's Web UI (for example, the built-in terminal, or VSCode Web),
+- Between workspace agents and `coderd`,
+- Between Coder [workspace proxies](../admin/workspace-proxies.md) and `coderd`.
+
+Any issues causing failures to establish WebSocket connections will result in **severe** impairment of functionality for
+users. To validate this functionality, Coder will periodically attempt to establish a WebSocket connection with itself
+using the configured [Access URL](#access-url), send a message over the connection, and attempt to read back that same message.
+
+### <a name="EWS01">EWS01: Failed to establish a Websocket Connection</a>
```
Comment (mtojek) [inline docs/admin/healthcheck.md:134]: nit: connection (lowercase)

File: docs/admin/healthcheck.md [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+
+**Problem:** Coder was unable to establish a WebSocket connection over its own Access URL.
+
+**Solution:** There are multiple possible causes of this problem:
+
+1. Ensure that Coder's configured Access URL is accessible from the server running Coder, using standard
```
Comment (mtojek) [inline docs/admin/healthcheck.md:140]: nit: Access URL is accessible (tautological?)

File: docs/admin/healthcheck.md [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+troubleshooting tools like `curl`.
+1. Ensure that any reverse proxy that is sitting in front of Coder's configured access URL is not stripping the HTTP header `Upgrade: websocket`.
+
+### <a name="EWS02">EWS02: Failed to echo a WebSocket message</a>
+
+**Problem:** Coder was able to establish a WebSocket connection, but was unable to write a message.
+
+**Solution:** There are multiple possible causes of this problem:
+
+1. Validate that any reverse proxy servers in front of Coder's configured access URL are not prematurely closing the connection.
```
Comment (mtojek) [inline docs/admin/healthcheck.md:150]: 2. Network stability?
3. Could the message size be an issue?

File: docs/admin/healthcheck.md (hunk 12 above)
Comment (johnstcn) [reply docs/admin/healthcheck.md]: I'm not sure of the right command here, so I'll just do a curl for now.

File: docs/admin/healthcheck.md (hunk 15 above)
Comment (mtojek) [reply docs/admin/healthcheck.md]: I would suggest checking policies related to firewalls, reverse proxies, VPNs, etc. Maybe it is "well known" that the corp network does not play well with long websocket connections.

Comment (johnstcn) [reply docs/admin/healthcheck.md]: added

**Review by mtojek — APPROVED** [review]
👍

File: docs/admin/healthcheck.md [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+
+This is not necessarily a fatal error, but a possible indication of a
+misconfigured reverse HTTP proxy. Additionally, while workspace users should
+still be able to reach their workspaces, connection performance may be degraded.
+
+> [!NOTE] This may also be shown if you have
+> [forced websocket connections for DERP](../cli/server.md#--derp-force-websockets).
+
+**Solution:** ensure that any configured reverse proxy does not strip the
+`Upgrade: derp` header.
```
Comment (spikecurtis) [inline docs/admin/healthcheck.md:138]: ensure that any proxies you use allow connection upgrade with the `Upgrade: derp` header.

File: docs/admin/healthcheck.md [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+
+### <a name="EDERP02">EDERP02: One or more DERP nodes are unhealthy</a>
+
+**Problem:** This is shown if Coder is unable to reach one or more configured
+DERP servers. Clients will fall back to use the remaining DERP servers, but
+performance may be impacted for clients closest to the unhealthy DERP server.
+
+**Solution:** Ensure that the DERP server is available and reachable over the
+network on port 443, for example:
```
Comment (spikecurtis) [inline docs/admin/healthcheck.md:147]: Ensure that the DERP server is available and reachable over the network, for example:

File: docs/admin/healthcheck.md [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+
+```shell
+curl -v "https://coder.company.com:443/derp"
```
Comment (spikecurtis) [inline docs/admin/healthcheck.md:150]: :443 is redundant if you use https://

File: docs/admin/healthcheck.md [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+**Problem:** Coder was unable to establish a WebSocket connection over its own
+Access URL.
+
+**Solution:** There are multiple possible causes of this problem:
+
+1. Ensure that Coder's configured Access URL can be reached from the server
+   running Coder, using standard troubleshooting tools like `curl`:
+
+   ```shell
+   curl -v "https://coder.company.com:443/"
```
Comment (spikecurtis) [inline docs/admin/healthcheck.md:184]: :443 is redundant with https://

File: docs/admin/healthcheck.md [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+   ```
+
+2. Ensure that any reverse proxy that is sitting in front of Coder's configured
+   access URL is not stripping the HTTP header `Upgrade: websocket`.
```
Comment (spikecurtis) [inline docs/admin/healthcheck.md:188]: Ensure that any reverse proxy that is serving Coder's configured access URL allows connection upgrade with the header `Upgrade: websocket`

File: docs/admin/healthcheck.md [hunk 21]
```diff
@@ -0,0 +1,248 @@
...
+
+**Solution:** Update the workspace proxy to match the currently running version
+of Coder.
+
+### <a name="EWP04">EWP04: One or more Workspace Proxies Unhealthy</a>
+
+**Problem:** One or more workspace proxies are not reachable.
+
+**Solution:** Ensure that Coder can establish a connection to the configured
+workspace proxies on port 443.
```
Comment (spikecurtis) [inline docs/admin/healthcheck.md:240]: Ensure that Coder can establish a connection to the configured
workspace proxies

**Review by spikecurtis — COMMENTED** [review]
Suggestions inline

File: docs/admin/healthcheck.md (hunk 21 above)
Comment (johnstcn) [reply docs/admin/healthcheck.md:240]: :+1:

File: docs/admin/healthcheck.md (hunk 20 above)
Comment (johnstcn) [reply docs/admin/healthcheck.md:188]: :+1:

File: docs/admin/healthcheck.md (hunk 16 above)
Comment (johnstcn) [reply docs/admin/healthcheck.md:138]: :+1:

**Pull request merged**